### DISCUSSION TITLE
[Bug修复指南]对于旧内核KernelSU不工作以及模块功能失效问题的部分解决方案

### DISCUSSION THREAD

**Discussion by Coconutat** [tiann/KernelSU#956] [Show and tell]
# 对于旧内核KernelSU不工作以及模块功能失效问题的部分解决方案
## 1.前言：  
**此指南献给还在用旧设备但没有自定义ROM的朋友。**  
针对的内核版本和安卓版本(更老的版本我没有设备，所以不一定只兼容下列)：    
内核版本：4.4，4.9。  
安卓版本：Android 8(O)，Android 9(P)。  
必要前置阅读：[手动集成KernelSU](https://kernelsu.org/zh_CN/guide/how-to-integrate-for-non-gki.html)   
提示1：如果KPROBES工作不正常，**你应该放弃KPROBES**，而不是死磕KPROBES。**代码是死的，人是活的。**  
***
提示2：  
对于手动集成，您应当把KPROBES关闭，不然会导致重启和崩溃之类的bug。
关闭的例子，在defconfig找到如下内容：
```
CONFIG_KPROBES=y
CONFIG_HAVE_KPROBES=y
CONFIG_KPROBE_EVENTS=y
```
改为：  
```
# CONFIG_KPROBES is not set
# CONFIG_HAVE_KPROBES is not set
# CONFIG_KPROBE_EVENTS is not set
```  
***
  
***  
  
## 2.对部分Bug的修复：
***
**Bug 1：SU工作但是模块不工作**   
**解决方案：修改hook.c**  
代码位于：  
```
security/selinux/hooks.c  
```
修改的patch:[security: selinux: allow init exec (kernelsu:ksud) under nosuid](https://github.com/sticpaper/android_kernel_xiaomi_msm8998-ksu/commit/646d0c836053ffa8743519c39fcabaecb258f00b)  
***
**Bug 2:低于安卓10的设备KernelSU显示工作但是SU授权无法正常使用且模块页面显示不支持Overlay_FS**  
**此Bug已经在v0.6.9修复。**
***
过时内容：
> **先声明**：您不应当把接下来的内容视为最终解决方案，因为不安全且也只是自欺欺人。  
因为KernelSU不兼容安卓10以下设备，所以无法正确处理低于安卓10的init。那么结果即使就会遭到SELinux的拦截。  
然而此时设备也无法通过adb命令切换SELinux状态，所以只能修改selinxfs.c的代码让其永远强制Permissive。  
这样的代价就是极其容易被检测和极大降低安全性。  
> 代码位于：  
```
security/selinux/selinuxfs.c   
```
> 修改的patch:[security:selinux:selinuxfs.c:Make SELinux Mode Always in permissive mode](https://github.com/Coconutat/android_kernel_huawei_florida/commit/2bfaa7fa742119062eeff38c37705c55a3fcd72b)  
***
***
其它的bug有待继续收集...

**Comment by longhuan1999** [comment]
对于Bug 2:

1. 安卓9 的 init 的 path 是 /init ；
2. init 是否进入 second_stage 通过 init 环境变量 "INIT_SECOND_STAGE" 判断。

已合并修复commit ：https://github.com/tiann/KernelSU/commit/54ee400dc5ccb521de761b4dfdff4613d36262e0

**Reply by Coconutat** [reply]
尝试给KernelSU提交PR，这样大家都可以受益

**Reply by dabao1955** [reply]
再提交pr之前，应该先考虑这是否会影响其他内核版本的ksu正常工作。

**Reply by Coconutat** [reply]
我在测试

**Reply by longhuan1999** [reply]
确实，所以我暂时还没有打算提交pr。尤其是以下修改是根据我手机的 dmesg 日志中的输出做出修改的，其实我对Android和C++是不太了解的。希望大家能够多多测试和审查我做出的修改
```
if (!strcmp(first_arg, "second_stage") || !strcmp(first_arg, "subcontext")) {
```

**Reply by longhuan1999** [reply]
我修改解决方案，安卓 9 的 init 进入第二阶段的判断改成了对 init 环境变量"INIT_SECOND_STAGE"值的判断。明天我会整理一下，提交pr

**Reply by Coconutat** [reply]
Good！

**Reply by Coconutat** [reply]
```
strncpy_from_user
```
这个参数我觉得你不应该直接使用。因为它不存在于高版本内核。weishu提交了一个commit用来解决这个问题，参考：https://github.com/tiann/KernelSU/commit/da959b4e17ce7f6ec48cd3da4d4bca0250310dc4

**Reply by Coconutat** [reply]
应该是直接用：
```
ksu_strncpy_from_user_nofault
```
就行了。代码会自动判断内核版本。

**Reply by longhuan1999** [reply]
谢谢，我会修改并测试一下

**Comment by hereblur** [comment]
When building with AOSP(crDroid in my case). Make sure  CONFIG_OVERLAY_FS  is set to “y” or CONFIG_KSU will be removed from actual .config.

**Reply by Separatee** [reply]
> When building with AOSP(crDroid in my case). Make sure CONFIG_OVERLAY_FS is set to “y” or CONFIG_KSU will be removed from actual .config.

btw, you should use script from offical repo
````
curl -LSs "https://raw.githubusercontent.com/tiann/KernelSU/main/kernel/setup.sh" | bash -s v0.9.5
````
it should be good, there is no need to change anything more, the script will help you to modify them.

**Comment by Pr0pHesyer** [comment]
建议将Bug 1：SU工作但是模块不工作相关内容直接写入文档而不是仅保留在discussion内

**Reply by Coconutat** [reply]
现在看似乎没有必要。1.0.0马上就要来了，non-GKI相关代码会被移除。到时候这些也用不上了。

**Comment by Yan233th** [comment]
赞
